drain and fill channel plugin

m-schmoock opened the pull request:
This plugin offers three new commands: `drain`, `fill` and `setbalance`.
It works similar to the `rebalance` plugin but with some key differences:

- Does not need a second `scid` parameter but figures out the other one by routing.
- Can have chunked payments that increases routing probability but can result in partial completion. If `cunks` is not set, number of required chunks is autodetected.
- Does not consume or calculate msat amounts, but uses percentages. Default `100` for `drain` and `fill` and `50` for `setbalance`.

Note: If draining 100%, the plugin guesses the correct HTLC fee by try and error. For new servers that report HTLC fee via exception (https://github.com/ElementsProject/lightning/pull/2691) it takes the exact amount.

Issues:
- Can run into "route not found" errors afters several "drain"s on multiple channels. Reason yet unknown.

m-schmoock commented:
I removed the "work in progress" TAG, as the plugin is very usable aready and I dont have good clues about the remaining TODOs and optimizations.

I could need a review now... volunteers?

darosior commented on drain/drain.py:
```diff
@@ -0,0 +1,418 @@
+        spendable, receivable = spendable_from_scid(plugin, payload, payload['scid'])
+        spendable_bak = spendable
+        total = spendable + receivable
+        #current_percentage = 100 * int(spendable) / int(total)
```
Trailing comment ? :)

darosior commented on drain/drain.py:
```diff
@@ -0,0 +1,418 @@
+        retry_for: int=60, exemptfee: Millisatoshi=Millisatoshi(5000)):
+    """Draining channel liquidity with circular payments.
+
+    Percentage defaults to 100. Chunks deaults to 0 (auto-detect).
```
```suggestion
    Percentage defaults to 100. Chunks defaults to 0 (auto-detect).
```

m-schmoock commented on drain/README.md:
````diff
@@ -0,0 +1,92 @@
+```
+
+Alternatively, you can also symlink or copy the plugins executable to the
+`plugins` folder of your c-lightning installation, as all of the executables
````
ack

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,418 @@
+
+    try:
+        channel_funds = next(c for c in payload['funds'] if 'short_channel_id' in c and c['short_channel_id'] == scid)
+    except StopIteration as e:
```
+1 moar nits, keep them coming ;)

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,418 @@
+    if their < their_reserve:
+        their_reserve = their
+
+    #spendable = our - our_reserve
```
yup

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,418 @@
+
+def find_worst_channel(route):
+    if len(route) < 4:
+        return None
```
hmmm, good point. this one was copy&pasta from @gallizoltan rebalance plugin. I think this was done to prevent removing/excluding too many hops if route is already too short or something.

Also, it does not seem to be related to the remaining drain 'Could not find a route' bug.

m-schmoock replied:
Hm, the `start_id` is 2, meaning it is the third entry in the route...

@gallizoltan can you elaborate on this version of the `find_worst_channel` function maybe (rebalance has the same) ?

gallizoltan replied:
Okay, it is partly because of historical reasons, and maybe obsolete.

When I wrote this I assumed we need at least 3 channels/hops for a rebalance:
1) The outgoing channel to the outgoing node
2) A middle channel to the incoming node
3) The incoming channel to my node

Number 1 and 3 are fixed. We could only exclude number 2.
But when I wrote this, there was no `setchannelfee` function. A node applied the same fee for all its channels. So there was no point to exclude channel number 2, because its fee set by the outgoing node, so we could not find a cheaper channel instead of that.
This is why `start_id` is 2: possible excludes starts with the third entry in the route.

m-schmoock commented:
> I also got an error when trying `drain` :
> 
> ```
> Error while processing drain: TypeError('__int__ returned non-int (type float)')
> ```

Which version of pylightnting are you using. Also can you give me a stacktrace or something on this? Would be very helpful to fix this...

m-schmoock commented:
@darosior 
> I also got an error when trying `drain` :
> 
> ```
> Error while processing drain: TypeError('__int__ returned non-int (type float)')
> ```

Which version of pylightnting are you using. Also can you give me a stacktrace or something on this? Would be very helpful to fix this...

darosior commented:
@m-schmoock sorry for the delayed answer.

> Which version of pylightnting are you using.

Last released one (0.7 I think), should I try with master's one ?

> Also can you give me a stacktrace or something on this?

Actually I don't have any. I tried to quickly go back through the execution but did not found from where it comes (didn't search that much though).

darosior commented:
Hi @m-schmoock, at which stage are you with this ? Is still work in progress because of the TODOs, or are there still bugs ? If this is only about the TODOs maybe we can merge it and other contributors could help improve this (definitely wanted :-)) plugin ?

m-schmoock commented:
Hi @darosior , kinda glad you asked! 

The plug-in works almost like a charm. The reason it is not merged yet is, that it raises a reproducible 'route not found' error after doing several 'drain' on varoius channels in a row.  I'm not 100% sure if this is caused by the plug-in or lightningd. If you then stop, remove gossip store, restart, wait and retry, the issue will reappear only after doing some drains again. If you have time and ambitions to debug it feel free. Also, this does not seem to happen on 'fill' which is why I am more unsure what causes it.

F.y.i. I will have more time on from November, since I reduced my regular job to just 25hrs a week :). This was going to be my first task anyway.

Technically and code wise, it is in a mergable state beside that.

darosior commented:
> The plug-in works almost like a charm. The reason it is not merged yet is, that it raises a reproducible 'route not found' error after doing several 'drain' on varoius channels in a row. I'm not 100% sure if this is caused by the plug-in or lightningd. If you then stop, remove gossip store, restart, wait and retry, the issue will reappear only after doing some drains again. If you have time and ambitions to debug it feel free. Also, this does not seem to happen on 'fill' which is why I am more unsure what causes it.

Hmmm interesting, we might leave this as an open issue ? Don't know what the repo captain thinks about that @cdecker :-) ?

> F.y.i. I will have more time on from November, since I reduced my regular job to just 25hrs a week :). This was going to be my first task anyway.

Great ! To work on Lightning ?

darosior commented:
Btw if you think it's mergeable you should remove the WIP tag

m-schmoock commented:
@darosior 

I removed the WIP flag, still Im going to fix this one way or another, now that we talked about this issues it is at least known. If you are fine we can merge, I lost my commit access because I have been inactive for a while, so we need someone else to merge it.

> > F.y.i. I will have more time on from November, since I reduced my regular job to just 25hrs a week :). This was going to be my first task anyway.

> Great ! To work on Lightning ?

Yes !

darosior commented:
> If you are fine we can merge, I lost my commit access because I have been inactive for a while, so we need someone else to merge it.

I don't have one so we'll have to wait for ackbot ;-)

m-schmoock commented on drain/requirements.txt:
```diff
@@ -0,0 +1 @@
+pylightning==0.0.7
```
thx. will go for ` >= 0.0.7.3 `

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+        if payload['command'] == 'drain':
+            r = plugin.rpc.getroute(my_id, payload['amount'], riskfactor=0,
+                    cltv=9, fromid=peer_id, fuzzpercent=0, exclude=excludes)
+            route_out = {'id': peer_id, 'channel': payload['scid'], 'direction': int(not my_id < peer_id)}
```
can change, it was "more readable" to me, dunno why :D

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+    target = total * (0.01 * (100 - payload['percentage']))
+
+    if cmd == "drain":
+        if target >= spendable:
```
Hm, because if they are equal the next line `amount = spendable - target` will set 0 amount which will raise errors later on..

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+    # check parameters
+    if command != 'drain' and command != 'fill':
+        raise RpcError(payload['command'], payload, {'message': 'Invalid command. Must be drain or fill.'})
+    if percentage < 0 or percentage > 100:
```
We can change back, once the plugin converts a "fill 0%" to a "drain 100%". Currently it will trow an error as the target is above/below spendable/receivable.

darosior commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+    target = total * (0.01 * (100 - payload['percentage']))
+
+    if cmd == "drain":
+        if target >= spendable:
```
FWIW after testing if I drain a channel at exactly 50% then these conditions are mutually exclusive ;-)

Ex:
```bash
$ lnreg1 drain 366x1x0 50 50
[
   "987762msat sent over 3 hops to drain 987740msat [1/50]",
   "997392msat sent over 3 hops to drain 997370msat [2/50]",
   "1007783msat sent over 3 hops to drain 1007759msat [3/50]",
   ....

$ lnreg1 drain 366x1x0 1 20
"Error while processing drain: RpcError(\"RPC call failed: ..... error: {'message': 'channel already below target value'}\")"
$ lnreg1 fill 366x1x0 1 20
"Error while processing fill: RpcError(\"RPC call failed: .......  error: {'message': 'channel already above target value'}\")"
```
```

darosior replied:
What's fun is that rebalanced all the regtest btw ^^

darosior replied:
Hmm think I'll have to test on the only viable test network : mainnet

darosior commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+                excludes.append(erring_channel + '/' + str(erring_direction))
+
+
+def execute(command: str, scid: str, percentage: int,
```
Trying this for real, I think it could be great if it could be a float as you may often need to not drain / fill a significant percentage of the channel funds

m-schmoock replied:
thx, addressed by 26e8717

darosior commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+    if cmd == "drain":
+        if target >= spendable:
+            raise RpcError(payload['command'], payload, {'message': 'channel already below target value'})
+        amount = spendable - target
```
Trying this some more I think I don't get the logic (or don't understand the params) :
I tried to drain some sats from a channel with `293600000`msat (`287728000` spendable), setting the percentage to 30 I effectively drained `33807374`msat, which is about 11%of the channel capacity:
```python
>>> 33807374 / 287728000 * 100
11.749768531390758
>>> 33807374 / 293600000 * 100
11.514773160762942
```
Hmm trying this again I then got an error that the target was above the capacity (or below to fill), even with 1% as parameter ! Is this the error you were talking about ?

Back to the highlighted lines I think you compute the target as the final channel capacity:
```python
target = total * (0.01 * (100 - payload['percentage']))
```
Then you compare it to the spendable amount
```python
if target >= spendable:
```
Which will likely be inferior.
I think that the target should be "how much do I want to drain" and thus the target be :
```python
target = total * payload['percentage'] / 100
```
Or 
```python
    # The amount to either fill or drain
    amount = total * payload['percentage'] / 100

    if cmd == "drain":
        if amount >= spendable:
             # Hmm problem here
    elif cmd == "fill":
        if amount >= receivable:
             # Hmmmm
```

But maybe I'm just misunderstanding and misusing the whole thing

m-schmoock replied:
Yes, the percentage amount is the resulting final channel capacity. The error can be reproduced when quickly draining and filling several channels (i.e. 100% default) until the error is raised.

m-schmoock replied:
Note: I edited my last response to be more precise.

darosior replied:
After more testing I think that's the latter case.. :
> But maybe I'm just misunderstanding and misusing the whole thing

darosior replied:
But I still don't understand why I could not drain (or fill) 1% of a channel that is completely balanced (nor the target logic)..

m-schmoock replied:
That's easy, a balanced channel is by definition 50% filled and drained 'at the same time'. You cannot be able to 'drain' to 1% as that would equate to a fill 99% (but command inversion is a Todo and not implemented yet.). Equally a fill to 1% is impossible as it has already 50% which is greater than 1%.

The percentage do not related to current balance but to target amount.

m-schmoock replied:
Do you think we should implement command handling differently, if yes how exactly?

Current pros:
- do not need to know current percentage l
- no amount calculation needed
- Able to rebalance by using drain or fill 50%

Cons:
- counter intuitive? :D
- 'not often possible' to do stuff like drain 1% as that would require to have the channel above 99%.

darosior replied:
Thanks for the explanation !

> Do you think we should implement command handling differently, if yes how exactly?

It was counter intuitive *to me*, maybe because I haven't read "mastering english" yet :smile: 
However I'd argue to maybe change the command name ? According to my understanding, `drain <amount> <scid>` implies `I want to drain <amount>% of sats from <scid>` : actually I tried to use it as a rebalance. Maybe `set_ours <percent>` `set_them <percent>`, they can still be additional commands though.

>  I think from a UX requirements perspective it is important to be able to get a channel balance full or empty (hence 100% default) or some to fractional value without doing math calculations. It is not a useful feature to send 1% or X% of a channel elsewhere.

I totally agree. Just asking as many questions as possible to make this as perfect as possible ;-) Next step (once merged) is the htlc fee finding optimisation

m-schmoock replied:
@darosior can you come to irc? Or can we have a quick talk somewhere?

darosior replied:
Yep I pinged you on IRC

darosior replied:
> How about this:

    drain X will send X% total (or current?) channel balance away to some other channel, if possible.
    fill X will receive X% total (or current?) channel balance from some other channel, if possible.
    reach X will try to achieve X% our capacity, if possible. This command will actually do the auto conversion (drain/fill) if channel is above or below.

Sounds great

darosior replied:
Ok so I have a channel with 60% theirs, 40% ours. I think the wording is misleading as if I want to drain, say 30 percent of the channel (so that the balance results in 90% theirs, 10% ours), I dont
```
cli drain <scid> 30
```
But
```
cli drain <scid> 90
```
So effectively passing to the drain command the expected percent on `their` side. That's why I proposed to change the wording to :
```
cli set_theirs <scid> 90
```
And, if we keep the `drain` command, to pass it the expected percent to __drain from__ my side.

The same goes for `fill` :
Now that they effectively have 90% of the channel capacity, and I want my side to have 90% capacity (so passing from `ours: 10%; theirs: 90%` to `ours: 90%; theirs: 10%`) I would expect to drain __from them__ 80% of the capacity (hence `fill` 80% on my side) but in order to do that with the current wording I have to :
```
cli fill <scid> 90
```
That's why I propose to rename it to 
```
cli set_ours <scid> 80
```

m-schmoock replied:
will be address by implementing this TODO 8d555a46dbd6dba32485ce4ea5a78955bae224ff

m-schmoock commented:
I updated the documentation a bit to make clear how it is currently supposed to work.

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+            r['amount_msat'] = amount_iter
+
+
+def get_channel(plugin, payload, peer_id, scid, check_state: bool=False):
```
thx. this was historic code. removed the check_state parameter all together and to the check or raise error.

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+    if percentage < 0 or percentage > 100:
+        raise RpcError(payload['command'], payload, {'message': 'Percentage must be between 0 and 100'})
+    if chunks < 0:
+        raise RpcError(payload['command'], payload, {'message': 'Negative chunks do not make sense. Try a positive value or use 0 (default) for auto-detection.'})
```
Chunk size is problematic from a UX, as it would require to do some math before invoking the operation. The chunk parameter is auto detected per default and should only be touched when automatic fails, as using more chunks than required increases fees...

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+    channels = plugin.rpc.listchannels(short_channel_id).get('channels')
+    for ch in channels:
+        if ch['source'] == my_id:
+            return ch['destination']
```
address by d9c462a

m-schmoock commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+    plugin.log("Plugin drain.py initialized")
+
+
+plugin.add_option('cltv-final', 10, 'Number of blocks for final CheckLockTimeVerify expiry')
```
Yes,

this one is copy&paste from rebalance and sendinvoiceless plugins.
It declares the `cltv-final` option the plugin is going to be using and also reads out lightningd's  `cltv-final` value in the `init` method. Maybe this is duplicated and we should always use `plugin.rpc.listconfigs().get('cltv-final')` but this increases RPC load.

So unless there is no other reason than duplication, I think we can keep it.

m-schmoock replied:
@darosior okay?

darosior replied:
> @darosior okay?

Yeah, tbh I don't use plugin options that much. But it's interesting that we actually can override `lightningd` options.

cdecker commented on drain/drain.py:
```diff
@@ -0,0 +1,417 @@
+#  - fix: occasionally strange route errors
+#  - feat: set HTLC_FEE MIN/MAX/STP by feerate
+#  - feat: convert a 'drain 0%' to a 'fill 100%' and vice versa instead of throwing error
+#  - chore: reconsider use of listchannels
```
Might be worth putting these in a markdown README in the subdirectory. Makes it clearer that there are some TODOs and allows you to tick them off visibly :-)

m-schmoock replied:
I will get this done in the next two weeks, I am about to have more time for all this... promise :D

m-schmoock replied:
@cdecker  I want to change the meaning of `drain` and `fill` to not mean 'total target percentge' but rather just 'total percentage', as it caused confusions for @darosior for being unintuitive. Ignore over draining/filling (replacing it by whats possible). And add a `setbalance` command that does all the auto stuff.

Regarding the routing errors, I will have to check. Maybe we want this merged first.

m-schmoock replied:
resolved by c8ac5cf `drain: document setbalance and todo`

cdecker commented:
Sorry for the delay, I seem to have missed the non-WIP window :wink: 

I'm quite happy to merge the plugin even if it has a edge-case that may fail some times. fwiw I have a c-lightning PR pending that should allow us to start testing plugins, so pinning down the issue and reproducing it will become easier: https://github.com/ElementsProject/lightning/pull/3218

darosior commented on drain/README.md:
````diff
@@ -0,0 +1,113 @@
+lightning-cli fill scid [percentage] [chunks] [maxfeepercent] [retry_for] [exemptfee]
+```
+
+Another useful command is the `setbalance` that will fill up or drain your side
````
:tada: ! :grin:

m-schmoock commented:
Hi,

After my first two days of my new 'Altersteilzeit' I managed to get this ready.

@darosior  I added the `setbalance` command and rewrote the `drain` `fill` to behave more intuitive. thanks for the feedback. The new command works like a charm.
@cdecker  I finally removed the WIP tag. Can we make final reviews and merge?

cheers,
Michael

m-schmoock commented:
You guys also might want to test this, but for my setup and testnet it works smoothly

darosior commented:
> testnet and mainnet it works smoothly.

So you don't have the spontaneous errors you were talking about ?

m-schmoock commented:
From time to time, which is why I think it's not plug-in but daemon related.
